rust/noosphere-sphere/src/sync/gateway.rs

```diff
@@ -151,73 +202,100 @@ where
             local_sphere_old_base,
             local_sphere_new_base,
         ) {
+            // History diverged, so rebase our local changes on the newly received branch
             (Some(current_tip), Some(old_base), Some(new_base)) => {
                 println!("Syncing received local sphere revisions...");
                 let new_tip = Sphere::at(current_tip, context.db())
-                    .try_sync(
+                    .sync(
                         &old_base,
                         &new_base,
                         &context.author().key,
                         context.author().authorization.as_ref(),
                     )
                     .await?;
 
-                context
-                    .db_mut()
-                    .set_version(&local_sphere_identity, &new_tip)
-                    .await?;
-
-                Some(new_tip)
+                new_tip
             }
+            // No diverged history, just new linear history based on our local tip
             (None, old_base, Some(new_base)) => {
                 println!("Hydrating received local sphere revisions...");
-                Sphere::try_hydrate_range(old_base.as_ref(), &new_base, context.db_mut()).await?;
+                Sphere::hydrate_range(old_base.as_ref(), &new_base, context.db_mut()).await?;
 
-                context
-                    .db_mut()
-                    .set_version(&local_sphere_identity, &new_base)
-                    .await?;
-
-                None
+                new_base
             }
-            _ => {
+            // No new history at all
+            (Some(current_tip), _, _) => {
                 println!("Nothing to sync!");
-                local_sphere_tip.cloned()
+                current_tip.clone()
+            }
+            // We should have local history but we don't!
+            _ => {
+                return Err(anyhow!("Missing local history for sphere after sync!"));
             }
         };
 
+        context
+            .db_mut()
+            .set_version(&local_sphere_identity, &local_sphere_tip)
+            .await?;
+
         debug!("Setting counterpart sphere version to {counterpart_sphere_tip}");
         context
             .db_mut()
             .set_version(counterpart_sphere_identity, &counterpart_sphere_tip)
             .await?;
 
-        Ok((local_sphere_tip, counterpart_sphere_tip))
+        Ok((local_sphere_tip, counterpart_sphere_tip, updated_names))
+    }
+
+    async fn adopt_names(
+        &self,
+        context: &mut H,
+        updated_names: BTreeMap<String, AddressIpld>,
+    ) -> Result<Option<Cid>> {
+        if updated_names.len() == 0 {
+            return Ok(None);
+        }
+        println!(
```

Comment:
No `println` in `noosphere` imo, this should be rendered in `orb` if needed.

There's plenty of existing println here though, let's kick off our rules to writing to stdout in a separate issue